File: yelp-dataload/src/main/java/org/abhijitsarkar/HttpReceiver.java
package org.abhijitsarkar;

import org.apache.spark.streaming.receiver.Receiver;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.zip.GZIPInputStream;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.apache.spark.storage.StorageLevel.MEMORY_AND_DISK;

/**
 * @author Abhijit Sarkar
 */
public class HttpReceiver extends Receiver<String> {
    private final String url;

    public HttpReceiver(String url) {
        super(MEMORY_AND_DISK());
        this.url = url;
    }

    @Override
    public void onStart() {
        new Thread(() -> receive()).start();
    }

    private void receive() {
        try {
            HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setAllowUserInteraction(false);
            conn.setInstanceFollowRedirects(true);
            conn.setRequestMethod("GET");
            conn.setReadTimeout(60 * 1000);

            InputStream gzipStream = new GZIPInputStream(conn.getInputStream());
            Reader decoder = new InputStreamReader(gzipStream, UTF_8);
            BufferedReader reader = new BufferedReader(decoder);

            String json = null;
            while (!isStopped() && (json = reader.readLine()) != null) {
                store(json);
            }
            reader.close();
            conn.disconnect();
        } catch (IOException e) {
            stop(e.getMessage(), e);
        }
    }

    @Override
    public void onStop() {

    }
}

File: yelp-dataload/src/main/scala/org/abhijitsarkar/HttpReceiver.scala
package org.abhijitsarkar

import java.net.{HttpURLConnection, URL}
import java.util.zip.GZIPInputStream

import org.apache.spark.storage.StorageLevel.MEMORY_AND_DISK
import org.apache.spark.streaming.receiver.Receiver

import scala.io.Source
import scala.util.{Failure, Success, Try}

/**
  * @author Abhijit Sarkar
  */
// https://spark.apache.org/docs/latest/streaming-custom-receivers.html
class HttpReceiver(val url: String) extends Receiver[String](MEMORY_AND_DISK) {
  // Must not block
  override def onStart(): Unit = new Thread(getClass.getSimpleName) {
    receive
  }.start

  private def receive = Try(new URL(url).openConnection().asInstanceOf[HttpURLConnection]) match {
    case Success(conn) => {
      conn.setAllowUserInteraction(false)
      conn.setInstanceFollowRedirects(true)
      conn.setRequestMethod("GET")
      conn.setReadTimeout(60 * 1000)

      val gzipStream = new GZIPInputStream(conn.getInputStream)

      Source.fromInputStream(gzipStream)
        .getLines
        .takeWhile(x => !isStopped && x != null)
        .foreach(store)

      conn.disconnect
    }
    case Failure(t) => stop(t.getMessage, t)
  }

  override def onStop(): Unit = {}
}

object HttpReceiver {
  def apply(url: String) = new HttpReceiver(url)
}
